oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java

```diff
@@ -281,14 +285,72 @@ private HttpResponse getMetadataResponse(String url) throws IOException {
     return response;
   }
 
-  /** Return whether code is running on Google Compute Engine. */
-  static boolean runningOnComputeEngine(
+  /**
+   * Implements an algorithm to detect whether the code is running on Google Compute Environment
+   * (GCE) or equivalent runtime. <a href="https://google.aip.dev/auth/4115">See AIP-4115 for more
+   * details</a> The algorithm consists of active and passive checks: <br>
+   * <b>Active:</b> to check that GCE Metadata service is present by sending a http request to send
+   * a request to {@code ComputeEngineCredentials.DEFAULT_METADATA_SERVER_URL}
+   *
+   * <p><b>Passive:</b> to check if SMBIOS variable is present and contains expected value. This
+   * step is platform specific:
+   *
+   * <p><b>For Linux:</b> check if the file "/sys/class/dmi/id/product_name" exists and contains a
+   * line that starts with Google.
+   *
+   * <p><b>For Windows:</b> to be implemented
+   *
+   * <p>This algorithm overall can be disabled with environment variable {@code
+   * DefaultCredentialsProvider.NO_GCE_CHECK_ENV_VAR} set to true. Returns {@code true} if currently
+   * running on Google Compute Environment (GCE) or equivalent runtime.
+   */
+  static synchronized boolean isOnGce(
       HttpTransportFactory transportFactory, DefaultCredentialsProvider provider) {
     // If the environment has requested that we do no GCE checks, return immediately.
     if (Boolean.parseBoolean(provider.getEnv(DefaultCredentialsProvider.NO_GCE_CHECK_ENV_VAR))) {
       return false;
     }
 
+    boolean result = pingComputeEngineMetadata(transportFactory, provider);
+
+    if (!result) {
+      result = checkStaticGceDetection(provider);
+    }
+
+    LOGGER.log(Level.FINE, "Failed to detect whether running on Google Compute Engine.");
+    return result;
+  }
+
+  @VisibleForTesting
+  static boolean checkProductNameOnLinux(BufferedReader reader) throws IOException {
+    String name = reader.readLine().trim();
+    return name.startsWith("Google");
+  }
+
+  @VisibleForTesting
+  static boolean checkStaticGceDetection(DefaultCredentialsProvider provider) {
+    String osName = provider.getOsName();
+    try {
+      if (osName.startsWith("linux")) {
```

Comment:
you can have a common method that gives an enum for the OS type by checking the name.

Reply:
added constants, enums drags more code without much value